OA-PASS/pass-docker
# Harvard pass-docker configuration

# Initial pass-docker configuration to support the Harvard pilot.

> When approved, this PR should probably be a **`merge`** commit, not `squash` or `rebase`

To invoke, use: `docker-compose -f harvard.yml <...>`
- **N.B.**: **`-f harvard.yml`**

Note that the *services* (e.g. `elasticsearch`, `fcrepo`, `indexer`, etc.) in the `harvard.yml` file are not name-spaced.  You can invoke `docker-compose down` without having to specify `-f harvard.yml` as long as the service names in `docker-compose.yml` and `harvard.yml` align.
## This PR provides
- a separate `docker-compose` file, `harvard.yml`.
- a separate environment file, `.harvard_env`
- adds `-harvard` to volume names, keeping them separate from JHU volume names.  Includes an initial Harvard assets image
- updates the `PI_ES_CONFIG` environment variable to use the latest ES configuration supporting [normalization of Harvard award numbers](https://github.com/OA-PASS/pass-data-model/pull/61)
- updated `assets` image complete with Harvard grant data, NIHMS submissions, Harvard `Repository` and `Policy` resources
- modifies the layout of some `pass-docker` images to support the separation of Harvard- and JHU- specific configuration
- Uses new _release_ versions of `authz` and `fcrepo` images containing PRs:
    - https://github.com/OA-PASS/pass-authz/pull/89
    - https://github.com/OA-PASS/pass-authz/pull/90
    - https://github.com/OA-PASS/pass-authz/issues/92 (issue for `User.affiliation` fix)
- Uses a _provisional_ `ember` image, containing PR:
   - https://github.com/OA-PASS/pass-ember/pull/1033 (Update model for `User.affiliation`)

Closes #235

## Related PRs:
- https://github.com/OA-PASS/pass-data-model/pull/61 (updated ES schema)
- https://github.com/OA-PASS/pass-data-model/pull/62 (updated JSON-LD context, related to capturing User affiliation)
- https://github.com/OA-PASS/java-fedora-client/pull/95 (update to latest JSON-LD context, related to capturing User affiliation)
- https://github.com/OA-PASS/metadata-schemas/pull/40 (expose Harvard CMF schemas on github.io)
- https://github.com/OA-PASS/pass-authz/pull/89 (configure auth domain with env var)
- https://github.com/OA-PASS/pass-authz/pull/90 (removes hopkins-isms from locator id identifier types)
- https://github.com/OA-PASS/pass-authz/pull/93 (update to latest JSON-LD context, capture User affiliation)
- https://github.com/OA-PASS/pass-policy-service/pull/17 (harvard policy)
- https://github.com/OA-PASS/pass-ember/pull/1033 (Ember model update to `User` and `Contributor`)

## Required releases
This PR incorporates releases of:
- pass data model version (i.e. JSON-LD context) 3.5
- pass java client v0.7.0
- pass-authz v0.5.0
- pass-policy-service v0.2.0
- a new release of deposit services (pending)
- a new ember release (pending https://github.com/OA-PASS/pass-ember/pull/1033)

## Testing Instructions
1. Shutdown and remove all existing containers
1. Remove all dangling volumes
1. Check out this pull request
1. Perform a `docker-compose -f harvard.yml pull`
1. Perform `docker-compose -f harvard.yml up -d`
1. Wait a bit for services to start

### Verify Assets Image
- Confirm there are no JHU Users; the following ES query should return no Users
`http://localhost:9200/pass/_search?q=@type:User+locatorIds:*hopkins*&pretty=true&default_operator=AND`
- Confirm there are no JHU Grants; the following ES query should return no Grants
`http://localhost:9200/pass/_search?q=@type:Grant+localKey:*hopkins*&pretty=true&default_operator=AND`
- Confirm there are no Users with a populated affiliation; the following ES query should return no Users
`http://localhost:9200/pass/_search?q=@type:User+_exists_:affiliation&pretty=true&default_operator=AND`
- Verify Submission count: 7135
`http://localhost:9200/pass/_search?q=@type:Submission&pretty=true&default_operator=AND`
- Verify Grant count: 3127
`http://localhost:9200/pass/_search?q=@type:Grant&pretty=true&default_operator=AND`
- Verify Policy resource for Harvard
`http://localhost:8080/fcrepo/rest/policies/f9/b6/01/25/f9b60125-662d-4e03-a7b0-eec0df2b50de`
    - Description text should contain HTML markup
- Verify Repository resource for Harvard
`http://fcrepo:8080/fcrepo/rest/repositories/93/c5/ff/37/93c5ff37-ca2b-4652-a2af-3b6794ff8790`
    - Agreement text should contain HTML markup
    - Links to schema should be to Harvard schema, and be valid links

### Verify test login
1. Login as User `faculty1`
    - Should have Harvard Grants and Submissions
    - Visit `/pass-user-service`
        - confirm `locatorIds` are populated with the following identifier types, and insure the values don't mention `johnshopkins.edu`:
            - employeeid
            - eppn
            - unique-id
        - confirm 'affiliation' is an array with one member: `FACULTY@harvard.edu`
        - confirm context uses version 3.5
1. Login as User `nih-user`
    - Should have Harvard Grants and Submissions
    - Visit `/pass-user-service`
        - confirm `locatorIds` are populated with the following identifier types, and insure the values don't mention `johnshopkins.edu`:
            - employeeid
            - eppn
            - unique-id
        - confirm 'affiliation' is an array with one member: `FACULTY@harvard.edu`
        - confirm context uses version 3.5

### Login and Perform Submission
1. Login as a test user using the normal PASS URL (https://pass.local)
1. Perform a Submission
    - Verify that Harvard DASH is a required repository
    - Verify Harvard First Author Affiliation is a required metadata element
1. View the Submission Details, where the submission to DASH ought to have a status of `submitted`
1. Open a new browser tab to approve the Submission in DSpace
1. Login to DSpace (http://pass.local:8181/xmlui)
    - Login link on upper right 
    - User dspace-admin@oa-pass.org password foobar
1. Click on "Submissions" on the right-hand navigation under "My Account"
1. Click on the title of the Submission, then click "Take"
    - May need to click the Take button agin
1. Then click "Approve"
    - This will accept the Submission into the DSpace archive
1. Go back to the PASS Submission Details, give things a minute or two, then click refresh. 
    - You should see the DASH Submission status be `accepted`
    - the provided link for the DASH Submission ought to open in DSpace.  The link will be something like:
        - `http://pass.local:8181/xmlui/handle/123456789/4`
1. Click on the link and view the full item metadata in DSpace.  Verify the following fields contain reasonable values:
    - `dc.relation.uri`	(this should be the PASS Submission URI)
    - `dash.funder.award`
    - `dash.funder.identifier`
    - `dash.funder.name`
    - `dash.affiliation.other`
